tests/serde-de.rs

```diff
@@ -6067,3 +6054,126 @@ fn from_str_should_ignore_encoding() {
         }
     );
 }
+
+/// Checks that deserializer is able to borrow data from the input
+mod borrow {
+    use super::*;
+
+    /// Struct that should borrow input to be able to deserialize successfully.
+    /// serde implicitly borrow `&str` and `&[u8]` even without `#[serde(borrow)]`
+    #[derive(Debug, Deserialize, PartialEq)]
+    struct BorrowedElement<'a> {
+        string: &'a str,
+    }
+
+    /// Struct that should borrow input to be able to deserialize successfully.
+    /// serde implicitly borrow `&str` and `&[u8]` even without `#[serde(borrow)]`
+    #[derive(Debug, Deserialize, PartialEq)]
+    struct BorrowedAttribute<'a> {
+        #[serde(rename = "@string")]
+        string: &'a str,
+    }
+
+    /// Deserialization of all XML's in that module expected to pass because
+    /// unescaping is not required, so deserialized `Borrowed*` types can hold
+    /// references to the original buffer with an XML text
+    mod non_escaped {
+        use super::*;
+        use pretty_assertions::assert_eq;
+
+        #[test]
+        fn top_level() {
+            let data: &str = from_str(r#"<root>without escape sequences</root>"#).unwrap();
+            assert_eq!(data, "without escape sequences",);
+        }
+
+        #[test]
+        fn element() {
+            let data: BorrowedElement = from_str(
+                r#"
+                <root>
+                    <string>without escape sequences</string>
+                </root>"#,
+            )
+            .unwrap();
+            assert_eq!(
+                data,
+                BorrowedElement {
+                    string: "without escape sequences",
+                }
+            );
+        }
+
+        #[test]
+        fn attribute() {
+            let data: BorrowedAttribute =
+                from_str(r#"<root string="without escape sequences"/>"#).unwrap();
+            assert_eq!(
+                data,
+                BorrowedAttribute {
+                    string: "without escape sequences",
+                }
+            );
+        }
+    }
+
+    /// Deserialization of all XML's in that module expected to fail because
+    /// values requires unescaping that will lead to allocating an internal
+    /// buffer by deserializer, but the `Borrowed*` types couldn't take ownership
+    /// on it.
```

Comment:
Is there any concept of opportunistic zero-copy with `Cow<&str>`?  As someone not terribly familiar with serde.

Reply:
If I understand you correctly, `Cow<str>` will borrow whenever possible i. e. when deserializer can provide data of underlying source by calling [`Visitor::visit_borrowed_str`](https://docs.rs/serde/latest/serde/de/trait.Visitor.html#method.visit_borrowed_str). The opposite case when the deserializer allocates `String` by itself and calls [`Visitor::visit_string`](https://docs.rs/serde/latest/serde/de/trait.Visitor.html#method.visit_string). The intermediate case when deserializer provides borrow to the data, that it owns, via [`Visitor::visit_str`](https://docs.rs/serde/latest/serde/de/trait.Visitor.html#method.visit_str).

In our case we can either provide access to the XML source by calling [`Visitor::visit_borrowed_str`](https://docs.rs/serde/latest/serde/de/trait.Visitor.html#method.visit_borrowed_str), or allocate a new `String` (because we in any case need to allocate to do unescaping and/or decoding) and give it to the `Visitor`. If XML in UTF-8 and does not contain escaped data or mixed text/CDATA or CDATA/CDATA content (after solving #474), the data will be borrowed